File: Neetcode/Graphs/countislands.cpp
#include <vector>
using namespace std;

class Solution
{
public:
  int numIslands(vector<vector<char>> &grid)
  {
    int ret = 0;
    for (int i = 0; i < grid.size(); i++)
    {
      for (int j = 0; j < grid[0].size(); j++)
      {
        if (grid[i][j] == '1')
        {
          ret += 1;
          dfs(grid, i, j);
        }
      }
    }
    return ret;
  }

private:
  void dfs(vector<vector<char>> &grid, int i, int j)
  {
    grid[i][j] = '0';
    if (i + 1 < grid.size() && grid[i + 1][j] == '1')
    {
      dfs(grid, i + 1, j);
    }
    if (i - 1 >= 0 && grid[i - 1][j] == '1')
    {
      dfs(grid, i - 1, j);
    }
    if (j + 1 < grid[0].size() && grid[i][j + 1] == '1')
    {
      dfs(grid, i, j + 1);
    }
    if (j - 1 >= 0 && grid[i][j - 1] == '1')
    {
      dfs(grid, i, j - 1);
    }
  }
};

File: Neetcode/Graphs/countislands.py
# https://neetcode.io/problems/count-number-of-islands
# https://leetcode.com/problems/number-of-islands/

from typing import List


class Solution:
    def numIslands(grid: List[List[str]]) -> int:
        row, col = len(grid), len(grid[0])
        ret = 0

        def dfs(r, c, grid):
            grid[r][c] = "0"
            if r < row - 1 and grid[r + 1][c] == "1":
                dfs(r + 1, c, grid)
            if r > 0 and grid[r - 1][c] == "1":
                dfs(r - 1, c, grid)
            if c < col - 1 and grid[r][c + 1] == "1":
                dfs(r, c + 1, grid)
            if c > 0 and grid[r][c - 1] == "1":
                dfs(r, c - 1, grid)

        for r in range(row):
            for c in range(col):
                if grid[r][c] == "1":
                    dfs(r, c, grid)
                    ret += 1
        return ret
